picking form block from helix website

vdua opened the pull request:
Test URLs:
- Before: https://main--nedbank--vdua.hlx.live/en/personal/borrow/personal-loans-form
- After: https://form-block--nedbank--vdua.hlx.live/en/personal/borrow/personal-loans-form


# Update 1

The review comments mentioned in the PRs have been merged here to make it easier to review. 

* applying style from content has been removed
* validation is done on form.submit event.
* element creation has been made simpler
* formatters are loaded dynamically

vdua committed:
picking form block from helix website

vdua committed:
reorganizing code

vdua committed:
removing code not required for nedbank

vdua committed:
support for more fieldTypes and generating unique ids for the fields

vdua committed:
support for formatting in fields

trieloff commented on blocks/form/form.js:
```diff
@@ -0,0 +1,182 @@
+  const payload = constructPayload(form);
+  const resp = await fetch(form.dataset.action, {
+    method: 'POST',
+    cache: 'no-cache',
```
```suggestion
```

This is probably not needed, as `POST` requests never get cached.

trieloff commented on blocks/form/form.js:
```diff
@@ -0,0 +1,182 @@
+  const button = document.createElement('button');
+  button.textContent = fd.Label;
+  button.classList.add('button');
+  if (fd.Type === 'submit') {
+    button.addEventListener('click', async (event) => {
+      const form = button.closest('form');
+      if (form.checkValidity()) {
+        event.preventDefault();
+        button.setAttribute('disabled', '');
+        await submitForm(form);
+        const redirectTo = fd.Extra;
+        window.location.href = redirectTo;
+      }
+    });
```
Two issues with this:
1. the validity check should not happen upon button click, but upon form submission (there are other ways to submit a form that would not get detected here)
2. disabling the submit button can be a good idea, but should happen in the validation cade

Alternative to the validity check upon form submission, there could be a check upon entry, but it's important to get this right:
- if you validity check upon first entry, all required fields will be marked as invalid, which is annoying (let me get to them first)
- if you validity check upon entry, email fields will first be marked invalid, until you get to the domain part, so not ideal either

vdua replied:
I will update the main form block in the helix-website then

trieloff replied:
Update it here, and we merge `helix-website` later.

vdua replied:
This code has already been removed as per Satya's review comments since nedbank has no use case of submit.

vdua replied:
There are multiple patterns related to validation

1. validate on as you type (which is there in nedbank as well), but it has problems when there are multiple constraints
2. validate after the change (or blur) event (this works correctly) but should happen only if there is a change in the value. Just using tabs to iterate through the fields should not validate.
3. validate on submit (the submit event) and on click in the case when we are not using form submit (which is happening here; I am assuming form service doesn't accept `application/url-encoded` content type). 
Both the click and submit event have the same problem as `form.submit` API doesn't trigger any event.

> disabling the submit button can be a good idea, but should happen in the validation cade
@trieloff I did not get this part. disabling is required in this case since we are doing a `fetch` here which is asynchronous and can lead to double submit. It should happen just before submit, validation is synchronous.

vdua replied:
I have resolved the first part of it.

trieloff commented on blocks/form/form.js:
```diff
@@ -0,0 +1,182 @@
+  input.type = fd.Type;
+  input.id = fd.Field;
+  input.setAttribute('placeholder', fd.Placeholder);
+  if (fd.Mandatory === 'x') {
```
You have this same block over and over again. A better way would be to have a `markRequired([element, spec])` function that you can `map` over and apply to all form elements that can be required.

vdua replied:
I know, but I did not change anything while copying the original block, assuming that the original block has been already approved. The refactoring has been done in the second PR.

trieloff commented on blocks/form/form.js:
```diff
@@ -0,0 +1,182 @@
+  json.data.forEach((fd) => {
+    fd.Type = fd.Type || 'text';
+    const fieldWrapper = document.createElement('div');
+    const style = fd.Style ? ` form-${fd.Style}` : '';
```
What would this field contain? Having style definitions in the content is a 🚩

vdua replied:
This is the original form block in helix-website. I thought that is a pattern that is being followed. We are not using that for nedbank.

trieloff replied:
Ok, then let's take it out for now. We can fix the `helix-website` later.

vdua replied:
Removed it

trieloff reviewed (changes requested):
- more concise element creation
- better handling of validations
- unclear usage of `Style`

vdua committed:
incorporating review comments

- handling submit event on form instead of button click.
- removing reading style from excel.
- adding a state in the form to prevent double submit as well (in case of multiple submit)
- in case of multiple submit buttons each submit can define a redirect location.

vdua committed:
Merge branch 'form-block1' into form-block

vdua committed:
Merge branch 'form-block3' into form-block

vdua committed:
using foreach instead of map for setting values

vdua committed:
Dynamically loading formatters

- also fixing some minor issues in the code

vdua committed:
missed the await statement before fetching formatters

vdua commented on blocks/form/form.js:
```diff
@@ -0,0 +1,267 @@
+const formatFns = await (async function imports() {
```
Though await is supported at top level but not sure what its impact will be on performance

[1] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#browser_compatibility

vdua commented on blocks/form/form.js:
```diff
@@ -0,0 +1,267 @@
+      return (tagName === 'BUTTON' || tagName === 'INPUT') && type === 'submit';
+    };
+    if (valid) {
+      if (isSubmitButton(e.submitter)) {
```
on form submit the submitter is a button. This check is not needed.

vdua committed:
making a common function to wrap a widget to field

- submit event is triggered after validation, so no need for checking the validity separately
